// dmn-js-shared/lib/base/View.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * Base view API to be implemented by viewer editors.
 */
var View = function () {
  function View() {
    _classCallCheck(this, View);
  }

  _createClass(View, [{
    key: "attachTo",


    /**
     * Attach view to parentNode.
     *
     * @param  {Element} parentNode
     */
    value: function attachTo(parentNode) {}

    /**
     * Detach view.
     */

  }, {
    key: "detach",
    value: function detach() {}

    /**
     * Open view on the given element.
     *
     * @param  {View}   view
     * @param  {Function} [done]
     */

  }, {
    key: "open",
    value: function open(view) {
      var done = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : noop;

      done();
    }

    /**
     * Subscribe to the given event.
     *
     * @param  {String}         event
     * @param  {Array<Object>}  args
     */

  }, {
    key: "on",
    value: function on(event) {}

    /**
     * Unsubscribe from given event.
     *
     * @param  {String}         event
     * @param  {Array<Object>}  args
     */

  }, {
    key: "off",
    value: function off(event) {}

    /**
     * Clear view (optional method).
     */

  }, {
    key: "clear",
    value: function clear() {}

    /**
     * Destroy view (optional method).
     */

  }, {
    key: "destroy",
    value: function destroy() {}
  }]);

  return View;
}();

// helpers //////////////////////

export default View;
function noop() {}
