// dmn-js-literal-expression/lib/Viewer.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import BaseViewer from 'dmn-js-shared/lib/base/viewer/Viewer';

import { assign } from 'min-dash';

import { domify, remove as domRemove } from 'min-dom';

import CoreModule from './core';
import DecisionPropertiesModule from './features/decision-properties';
import LiteralExpressionPropertiesModule from './features/literal-expression-properties';
import PoweredByModule from './features/powered-by';
import TextareaModule from './features/textarea';
import ViewDrdModule from './features/view-drd';

var Viewer = function (_BaseViewer) {
  _inherits(Viewer, _BaseViewer);

  function Viewer() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, Viewer);

    var container = Viewer._createContainer();

    var _this = _possibleConstructorReturn(this, (Viewer.__proto__ || Object.getPrototypeOf(Viewer)).call(this, assign(options, {
      renderer: {
        container: container
      }
    })));

    _this._container = container;
    return _this;
  }

  _createClass(Viewer, [{
    key: 'open',
    value: function open(decision, done) {

      var err;

      // use try/catch to not swallow synchronous exceptions
      // that may be raised during model parsing
      try {

        if (this._decision) {

          // clear existing literal expression
          this.clear();

          // unmount first
          this.get('eventBus').fire('renderer.unmount');
        }

        // update literal expression
        this._decision = decision;

        // let others know about import
        this.get('eventBus').fire('import', decision);

        this.get('eventBus').fire('renderer.mount');
      } catch (e) {
        err = e;
      }

      // handle synchronously thrown exception
      return done(err);
    }

    /**
     * Initialize the literal expression, returning { modules: [], config }.
     *
     * @param  {Object} options
     *
     * @return {Object} init config
     */

  }, {
    key: '_init',
    value: function _init(options) {
      var modules = options.modules,
          additionalModules = options.additionalModules,
          config = _objectWithoutProperties(options, ['modules', 'additionalModules']);

      var baseModules = modules || this.getModules();
      var extraModules = additionalModules || [];
      var staticModules = [{
        viewer: ['value', this]
      }];

      var allModules = [].concat(_toConsumableArray(baseModules), _toConsumableArray(extraModules), staticModules);

      return {
        modules: allModules,
        config: config
      };
    }

    /**
     * Register an event listener
     *
     * Remove a previously added listener via {@link #off(event, callback)}.
     *
     * @param {String} event
     * @param {Number} [priority]
     * @param {Function} callback
     * @param {Object} [that]
     */

  }, {
    key: 'on',
    value: function on(event, priority, callback, target) {
      return this.get('eventBus').on(event, priority, callback, target);
    }

    /**
     * De-register an event listener
     *
     * @param {String} event
     * @param {Function} callback
     */

  }, {
    key: 'off',
    value: function off(event, callback) {
      this.get('eventBus').off(event, callback);
    }

    /**
     * Emit an event on the underlying {@link EventBus}
     *
     * @param  {String} type
     * @param  {Object} event
     *
     * @return {Object} event processing result (if any)
     */

  }, {
    key: '_emit',
    value: function _emit(type, event) {
      return this.get('eventBus').fire(type, event);
    }

    /**
     * Returns the currently displayed decision.
     *
     * @return {ModdleElement}
     */

  }, {
    key: 'getDecision',
    value: function getDecision() {
      return this._decision;
    }

    /**
     * Attach viewer to given parent node.
     *
     * @param  {Element} parentNode
     */

  }, {
    key: 'attachTo',
    value: function attachTo(parentNode) {

      if (!parentNode) {
        throw new Error('parentNode required');
      }

      // ensure we detach from the
      // previous, old parent
      this.detach();

      parentNode.appendChild(this._container);

      this._emit('attach', {});
    }

    /**
     * Detach viewer from parent node, if attached.
     */

  }, {
    key: 'detach',
    value: function detach() {

      var container = this._container,
          parentNode = container.parentNode;

      if (!parentNode) {
        return;
      }

      this._emit('detach', {});

      domRemove(container);
    }
  }, {
    key: 'destroy',
    value: function destroy() {
      _get(Viewer.prototype.__proto__ || Object.getPrototypeOf(Viewer.prototype), 'destroy', this).call(this);

      this.detach();
    }
  }, {
    key: 'getModules',
    value: function getModules() {
      return Viewer._getModules();
    }
  }], [{
    key: '_getModules',
    value: function _getModules() {
      return [CoreModule, DecisionPropertiesModule, LiteralExpressionPropertiesModule, PoweredByModule, TextareaModule, ViewDrdModule];
    }
  }, {
    key: '_createContainer',
    value: function _createContainer() {
      return domify('<div class="dmn-literal-expression-container"></div>');
    }
  }]);

  return Viewer;
}(BaseViewer);

export default Viewer;
