// dmn-js-decision-table/lib/features/decision-table-head/editor/components/InputCell.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

import { createVNode } from 'inferno';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import { Component } from 'inferno';

import { mixin } from 'table-js/lib/components';

import { ComponentWithSlots } from 'dmn-js-shared/lib/components/mixins';

var InputCell = function (_Component) {
  _inherits(InputCell, _Component);

  function InputCell(props, context) {
    _classCallCheck(this, InputCell);

    var _this = _possibleConstructorReturn(this, (InputCell.__proto__ || Object.getPrototypeOf(InputCell)).call(this, props, context));

    _this.onClick = function (event) {
      var input = _this.props.input;


      _this._eventBus.fire('input.edit', {
        event: event,
        input: input
      });
    };

    _this.onContextmenu = function (event) {
      var id = _this.props.input.id;


      _this._eventBus.fire('cell.contextmenu', {
        event: event,
        id: id
      });
    };

    _this.onElementsChanged = function () {
      _this.forceUpdate();
    };

    mixin(_this, ComponentWithSlots);
    return _this;
  }

  _createClass(InputCell, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var injector = this.context.injector;


      this._changeSupport = this.context.changeSupport;
      this._sheet = injector.get('sheet');
      this._eventBus = injector.get('eventBus');
      this._elementRegistry = injector.get('elementRegistry');

      var root = this._sheet.getRoot();

      var input = this.props.input;


      this._changeSupport.onElementsChanged(root.id, this.onElementsChanged);
      this._changeSupport.onElementsChanged(input.id, this.onElementsChanged);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      var root = this._sheet.getRoot();

      var input = this.props.input;


      this._changeSupport.offElementsChanged(root.id, this.onElementsChanged);
      this._changeSupport.offElementsChanged(input.id, this.onElementsChanged);
    }
  }, {
    key: 'render',
    value: function render() {
      var input = this.props.input;

      var inputExpression = input.inputExpression;


      var label = input.get('label');
      var inputVariable = input.get('inputVariable');

      var expressionLanguage = inputExpression.get('expressionLanguage') || 'FEEL';

      var showLanguageBadge = !label && expressionLanguage != 'FEEL';

      return createVNode(1, 'th', 'input-cell input-editor', [this.slotFills({
        type: 'cell-inner',
        context: {
          cellType: 'input-cell',
          col: this._elementRegistry.get(input.id)
        },
        col: input
      }), label ? createVNode(1, 'span', 'input-label', label, 0, {
        'title': 'Input Label'
      }) : createVNode(1, 'span', 'input-expression', inputExpression.text || '-', 0, {
        'title': 'Input Expression'
      }), inputVariable && createVNode(1, 'span', 'dms-badge dmn-variable-name input-variable', inputVariable, 0, {
        'title': 'Input Variable'
      }), showLanguageBadge && createVNode(1, 'span', 'dms-badge dmn-expression-language input-expression-language', [createVNode(1, 'span', 'dms-badge-icon dmn-icon-file-code'), createVNode(1, 'span', 'dms-badge-label', expressionLanguage, 0)], 4, {
        'title': 'Input Expression Language = ' + expressionLanguage
      })], 0, {
        'data-col-id': input.id,
        'onClick': this.onClick,
        'onContextmenu': this.onContextmenu
      });
    }
  }]);

  return InputCell;
}(Component);

export default InputCell;
