// dmn-js-shared/lib/base/viewer/core/components/ViewerComponent.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

import { createVNode, createComponentVNode } from 'inferno';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import { Component } from 'inferno';

var ViewerComponent = function (_Component) {
  _inherits(ViewerComponent, _Component);

  function ViewerComponent(props) {
    _classCallCheck(this, ViewerComponent);

    var _this = _possibleConstructorReturn(this, (ViewerComponent.__proto__ || Object.getPrototypeOf(ViewerComponent)).call(this, props));

    var injector = _this._injector = props.injector;

    _this._changeSupport = injector.get('changeSupport');
    _this._components = injector.get('components');
    _this._renderer = injector.get('renderer');
    return _this;
  }

  _createClass(ViewerComponent, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return {
        changeSupport: this._changeSupport,
        components: this._components,
        renderer: this._renderer,
        injector: this._injector
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var components = this._components.getComponents('viewer');

      return createVNode(1, 'div', 'viewer-container', components && components.map(function (Component, index) {
        return createComponentVNode(2, Component, null, index);
      }), 0);
    }
  }]);

  return ViewerComponent;
}(Component);

export default ViewerComponent;
